File: netlify/functions/film_opening.mjs
import prompt from "../../lib/assets/film-opening-prompt.json" with { type: "json" };
import { narrationLanguageName } from "./narration_language.mjs";

export function openingRequestBody(value) {
  const language = narrationLanguageName(value?.language);
  const characterName = validateCharacterName(value?.characterName);
  return {
    model: prompt.model,
    reasoning: { effort: "high" },
    // High reasoning effort consumes this same budget before the JSON output.
    max_output_tokens: 2400,
    store: false,
    provider: { sort: "latency" },
    instructions: prompt.instructions,
    input: `Create a new opening in ${language}. The protagonist name is ${JSON.stringify(characterName)}.`,
  };
}

export function validateCharacterName(value) {
  if (typeof value !== "string") throw new Error("Missing protagonist name.");
  if (/[\u0000-\u001f\u007f]/.test(value)) {
    throw new Error("Invalid protagonist name.");
  }
  const name = value.trim().replace(/\s+/g, " ");
  if (!name || name.length > 60) {
    throw new Error("Invalid protagonist name.");
  }
  return name;
}

export function parseOpening(value) {
  const limits = { title: 90, director: 70, narration: 900 };
  const opening = {};
  for (const [key, limit] of Object.entries(limits)) {
    const text = value?.[key];
    if (typeof text !== "string" || !text.trim() || text.length > limit) {
      throw new Error(`Invalid opening ${key}.`);
    }
    opening[key] = text.trim();
  }
  return opening;
}

export async function generateOpening(value, options) {
  const body = openingRequestBody(value);
  const response = await (options.fetchImpl ?? fetch)("https://openrouter.ai/api/v1/responses", {
    method: "POST",
    headers: {
      Authorization: `Bearer ${options.openRouterApiKey}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(body),
    signal: AbortSignal.timeout(28_000),
  });
  if (!response.ok) throw new Error(`Opening generation failed with HTTP ${response.status}.`);
  const result = await response.json();
  if (result.status === "incomplete" || result.error) {
    throw new Error("Opening generation did not complete.");
  }
  const text = result.output_text ?? (result.output ?? [])
    .flatMap((item) => item.content ?? [])
    .filter((part) => part.type === "output_text")
    .map((part) => part.text).join("");
  return parseOpening(JSON.parse(text));
}

File: lib/film_opening.dart
import 'dart:convert';

import 'package:flutter/services.dart';
import 'package:mcgee/narration_engine.dart';
import 'package:mcgee/openai.dart';

import 'narrator_profile.dart';

/// Fictional credits and scene-independent voiceover generated before capture.
final class FilmOpening {
  const FilmOpening({
    required this.title,
    required this.director,
    required this.narration,
  });

  factory FilmOpening.fromJson(Object? value) {
    if (value is! Map) {
      throw const FormatException('Opening must be a JSON object.');
    }
    String field(String name) {
      final text = value[name];
      if (text is! String || text.trim().isEmpty) {
        throw FormatException('Invalid opening $name.');
      }
      return text.trim();
    }

    return FilmOpening(
      title: field('title'),
      director: field('director'),
      narration: field('narration'),
    );
  }

  final String title;
  final String director;
  final String narration;
}

final class PreparedFilmOpening {
  const PreparedFilmOpening({required this.opening, required this.track});

  final FilmOpening opening;
  final AudioTrack track;
}

Future<FilmOpening> generateFilmOpening(
  OpenAiApi client,
  NarrationLanguage language, {
  required String characterName,
  NarratorProfile? profile,
}) async {
  final selectedProfile = profile ?? (await loadNarratorProfiles())['morgan-freeman']!;
  final config =
      jsonDecode(
            await rootBundle.loadString('lib/assets/film-opening-prompt.json'),
          )
          as Map<String, dynamic>;
  final response = await client
      .createResponse({
        'model': config['model'],
        'reasoning': {'effort': 'medium'},
        'max_output_tokens': 1200,
        'store': false,
        'instructions': '${config['instructions']}\n'
            '${selectedProfile.openingInstructions}',
        'input': 'Create a new opening in ${language.englishName}. The '
            'protagonist name is ${jsonEncode(characterName)}.',
      })
      .timeout(const Duration(seconds: 30));
  return _openingFromResponse(response);
}

FilmOpening _openingFromResponse(Map<String, Object?> response) {
  if (response['status'] == 'incomplete' || response['error'] != null) {
    throw const FormatException('Opening generation did not complete.');
  }
  final direct = response['output_text'];
  final output = response['output'];
  if (direct is String) return FilmOpening.fromJson(jsonDecode(direct));
  final text = StringBuffer();
  if (output is List) {
    for (final item in output.whereType<Map>()) {
      final content = item['content'];
      if (content is! List) continue;
      for (final part in content.whereType<Map>()) {
        if (part['type'] == 'output_text' && part['text'] is String) {
          text.write(part['text']);
        }
      }
    }
  }
  return FilmOpening.fromJson(jsonDecode(text.toString()));
}
